### ISSUE TITLE
shellcraft.mips.linux.execve is broken/behaving oddly

### ISSUE THREAD

**Issue by 0x27** [Gallopsled/pwntools#693]
This could be down to user error, however, the below log fairly well produces an example of the issue.

This issue was found while trying to work around this issue: https://github.com/Gallopsled/pwntools/issues/692

Expected Behaviour: i386 execve and mips execve both behave the same.
Actual Behaviour: Errors occur.

```
>>> from pwn import *
>>> path = '/bin/sh'
>>> argv = ['sh', '-c', 'echo Hello, $NAME; exit $STATUS']
>>> envp = {'NAME': 'zerocool', 'STATUS': 3}
>>> sc = shellcraft.i386.linux.execve(path, argv, envp)
>>> sc2 = shellcraft.mips.linux.execve(path, argv, envp)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 8, in execve
  File "/usr/local/lib/python2.7/dist-packages/mako/template.py", line 445, in render
    return runtime._render(self, self.callable_, args, data)
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 829, in _render
    **_kwargs_for_callable(callable_, data))
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 864, in _render_context
    _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 890, in _exec_template
    callable_(context, *args, **kwargs)
  File "/home/sky/.pwntools-cache/mako/mips/linux/execve.asm.py", line 32, in render_body
    __M_writer(unicode(syscall('SYS_execve', path, argv, envp)))
  File "<string>", line 8, in syscall
  File "/usr/local/lib/python2.7/dist-packages/mako/template.py", line 445, in render
    return runtime._render(self, self.callable_, args, data)
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 829, in _render
    **_kwargs_for_callable(callable_, data))
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 864, in _render_context
    _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
  File "/usr/local/lib/python2.7/dist-packages/mako/runtime.py", line 890, in _exec_template
    callable_(context, *args, **kwargs)
  File "/home/sky/.pwntools-cache/mako/mips/linux/syscall.asm.py", line 59, in render_body
    syscall_repr = syscall_repr % ', '.join(args)
TypeError: sequence item 1: expected string, list found
>>> 
```

**Comment by zachriggle** [comment]
The MIPS `execve` syscall is a dumb wrapper.  There is no existing infrastructure in MIPS to support for NULL-terminated argument lists (vs. `i386.pushstr_array`).

You'll notice that their behavior is consistent with the documentation for each function.

```
$ shellcraft i386.linux.execve -?
Execute a different process.

Attempts to perform some automatic detection of types.
Otherwise, the arguments behave as normal.

- If ``path`` is a string that is not a known register,
  it is pushed onto the stack.
- If ``argv`` is an array of strings, it is pushed onto
  the stack, and NULL-terminated.
- If ``envp`` is an dictionary of {string:string},
  it is pushed onto the stack, and NULL-terminated.
```

```
$ shellcraft mips.linux.execve -?
Invokes the syscall execve.  See 'man 2 execve' for more information.

Arguments:
    path(char): path
    argv(char): argv
    envp(char): envp
```

**Comment by zachriggle** [comment]
Fixed with f4a72da40e3e17090b793c866ac1c70b0bc4d173 and 6376d072660fb2250f48bd22629bbd7e3c61c758

```
$ shellcraft mips.linux.execve '/bin/sh' '["sh","-c","echo $FOO"]' "{'FOO':'Hello, world'}"  --run SILENT
Hello, world
```
